File: rust/src/process_socket.rs
use std::time::Duration;

use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine as _;
use futures_util::{SinkExt, StreamExt};
use http::header::{HeaderName, HeaderValue};
use serde_json::Value;
use tokio::net::TcpStream;
use tokio::time::timeout;
use tokio_tungstenite::tungstenite::client::IntoClientRequest;
use tokio_tungstenite::tungstenite::Message;
use tokio_tungstenite::{connect_async, MaybeTlsStream, WebSocketStream};

use crate::config::KEEPALIVE_PING_INTERVAL_SECS;
use crate::error::{Error, Result};

/// Streaming WebSocket connection to the sandbox process runtime.
pub struct ProcessSocket {
    socket: WebSocketStream<MaybeTlsStream<TcpStream>>,
}

impl ProcessSocket {
    /// Connect to a process runtime WebSocket path with bearer-token auth.
    pub async fn connect(base_url: &str, token: &str, path: &str) -> Result<Self> {
        let mut request = ws_url(base_url, path)?.into_client_request()?;
        request.headers_mut().insert(
            HeaderName::from_static("authorization"),
            HeaderValue::from_str(&format!("Bearer {token}"))?,
        );
        let (socket, _response) = connect_async(request).await?;
        Ok(Self { socket })
    }

    /// Send a JSON frame to the process runtime.
    pub async fn send_json(&mut self, payload: &Value) -> Result<()> {
        self.socket
            .send(Message::Text(payload.to_string().into()))
            .await?;
        Ok(())
    }

    /// Send stdin bytes encoded in the sandbox runtime protocol.
    pub async fn send_stdin(&mut self, data: impl AsRef<[u8]>) -> Result<()> {
        self.send_json(&stdin_payload(data)).await
    }

    /// Close stdin for the attached process.
    pub async fn close_stdin(&mut self) -> Result<()> {
        self.send_json(&close_stdin_payload()).await
    }

    /// Send a WebSocket ping frame.
    pub async fn send_ping(&mut self) -> Result<()> {
        self.socket
            .send(Message::Ping(b"watasu-sdk".to_vec().into()))
            .await?;
        Ok(())
    }

    /// Close the local WebSocket stream.
    pub async fn close(&mut self) -> Result<()> {
        self.socket.close(None).await?;
        Ok(())
    }

    /// Read the next JSON process frame.
    pub async fn next_frame(&mut self) -> Result<Option<Value>> {
        let idle = Duration::from_secs((KEEPALIVE_PING_INTERVAL_SECS / 2).max(1));

        loop {
            match timeout(idle, self.socket.next()).await {
                Ok(Some(Ok(message))) => match message {
                    Message::Text(text) => {
                        let frame: Value = serde_json::from_str(&text)?;
                        match frame.get("type").and_then(Value::as_str) {
                            Some("ready" | "pong") => continue,
                            Some("error") => {
                                let message = frame
                                    .get("message")
                                    .or_else(|| frame.get("code"))
                                    .and_then(Value::as_str)
                                    .unwrap_or("process error");
                                return Err(Error::Sandbox(message.to_string()));
                            }
                            _ => return Ok(Some(frame)),
                        }
                    }
                    Message::Binary(_) => {
                        return Err(Error::Sandbox(
                            "process websocket returned binary frame".into(),
                        ))
                    }
                    Message::Close(_) => return Ok(None),
                    Message::Ping(payload) => self.socket.send(Message::Pong(payload)).await?,
                    Message::Pong(_) => continue,
                    Message::Frame(_) => continue,
                },
                Ok(Some(Err(error))) => return Err(error.into()),
                Ok(None) => return Ok(None),
                Err(_elapsed) => {
                    self.send_ping().await?;
                    continue;
                }
            }
        }
    }

    /// Keepalive interval used by the SDK.
    pub fn keepalive_interval_secs(&self) -> u64 {
        KEEPALIVE_PING_INTERVAL_SECS
    }
}

/// Encode stdin bytes for the sandbox runtime protocol.
pub fn encode_runtime_data(data: impl AsRef<[u8]>) -> String {
    BASE64.encode(data)
}

pub(crate) fn stdin_payload(data: impl AsRef<[u8]>) -> Value {
    serde_json::json!({
        "type": "stdin",
        "data": encode_runtime_data(data)
    })
}

pub(crate) fn close_stdin_payload() -> Value {
    serde_json::json!({"type": "close_stdin"})
}

/// Decode base64 stdout/stderr frame data from the sandbox runtime protocol.
pub fn decode_runtime_data(value: &str) -> String {
    BASE64
        .decode(value)
        .map(|bytes| String::from_utf8_lossy(&bytes).into_owned())
        .unwrap_or_else(|_| value.to_string())
}

fn ws_url(base_url: &str, path: &str) -> Result<String> {
    let mut url = url::Url::parse(base_url)?;
    let scheme = match url.scheme() {
        "https" => "wss",
        "http" => "ws",
        other => other,
    }
    .to_string();
    url.set_scheme(&scheme)
        .map_err(|_| Error::Sandbox("invalid websocket scheme".into()))?;
    url.set_path(path.split('?').next().unwrap_or(path));
    if let Some(query) = path.split_once('?').map(|(_, query)| query) {
        url.set_query(Some(query));
    }
    Ok(url.to_string())
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::{close_stdin_payload, stdin_payload};

    #[test]
    fn process_input_payloads_match_runtime_protocol() {
        assert_eq!(
            stdin_payload("hi\n"),
            json!({"type": "stdin", "data": "aGkK"})
        );
        assert_eq!(close_stdin_payload(), json!({"type": "close_stdin"}));
    }
}

File: ts/src/processSocket.ts
import { Buffer } from 'node:buffer'

import WebSocket, { type RawData } from 'ws'

import { KEEPALIVE_PING_INTERVAL_SEC } from './connectionConfig.js'
import { SandboxError, TimeoutError } from './errors.js'

export type ProcessFrame = Record<string, unknown>

/** Streaming WebSocket connection to the sandbox process runtime. */
export class ProcessSocket implements AsyncIterable<ProcessFrame> {
  private ws?: WebSocket
  private queue: ProcessFrame[] = []
  private waiters: Array<(value: IteratorResult<ProcessFrame>) => void> = []
  private ackWaiters = new Map<string, Array<{ resolve: () => void; reject: (error: Error) => void; timer: ReturnType<typeof setTimeout> }>>()
  private closed = false
  private keepalive?: ReturnType<typeof setInterval>

  constructor(
    private readonly baseUrl: string,
    private readonly token: string,
    private readonly path: string,
    private readonly requestTimeoutMs = 60_000,
    private readonly headers: Record<string, string> = {}
  ) {}

  async connect(): Promise<this> {
    const ws = new WebSocket(wsUrl(this.baseUrl, this.path), {
      headers: { ...this.headers, Authorization: `Bearer ${this.token}` },
    })
    this.ws = ws

    ws.on('message', (data) => this.onMessage(data))
    ws.on('close', () => this.finish())
    ws.on('error', () => this.finish(new SandboxError('process websocket failed')))

    await new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => reject(new TimeoutError()), this.requestTimeoutMs)
      ws.once('open', () => {
        clearTimeout(timeout)
        resolve()
      })
      ws.once('error', () => {
        clearTimeout(timeout)
        reject(new SandboxError('process websocket failed to connect'))
      })
    })

    this.keepalive = setInterval(() => {
      if (this.ws?.readyState === WebSocket.OPEN) {
        this.ws.ping('watasu-sdk')
      }
    }, KEEPALIVE_PING_INTERVAL_SEC * 1000)
    return this
  }

  async sendJson(payload: ProcessFrame): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new SandboxError('process websocket is not connected')
    }
    await new Promise<void>((resolve, reject) => {
      this.ws!.send(JSON.stringify(payload), (error) => {
        if (error) reject(error)
        else resolve()
      })
    })
  }

  async sendStdin(data: string | Uint8Array): Promise<void> {
    const raw = typeof data === 'string' ? new TextEncoder().encode(data) : data
    const ack = this.waitForControlAck('stdin_ack')
    try {
      await this.sendJson({ type: 'stdin', data: base64Encode(raw) })
      await ack.promise
    } catch (error) {
      ack.cancel()
      throw error
    }
  }

  async closeStdin(): Promise<void> {
    const ack = this.waitForControlAck('close_stdin_ack')
    try {
      await this.sendJson({ type: 'close_stdin' })
      await ack.promise
    } catch (error) {
      ack.cancel()
      throw error
    }
  }

  close(): void {
    this.closed = true
    if (this.keepalive) clearInterval(this.keepalive)
    this.ws?.close()
    this.flushDone()
  }

  [Symbol.asyncIterator](): AsyncIterator<ProcessFrame> {
    return {
      next: () => this.next(),
    }
  }

  private next(): Promise<IteratorResult<ProcessFrame>> {
    if (this.queue.length > 0) {
      return Promise.resolve({ done: false, value: this.queue.shift()! })
    }
    if (this.closed) return Promise.resolve({ done: true, value: undefined })
    return new Promise((resolve) => this.waiters.push(resolve))
  }

  private onMessage(message: RawData): void {
    const text = rawDataToText(message)
    try {
      const frame = JSON.parse(text) as ProcessFrame
      if (frame.type === 'pong' || frame.type === 'ready') return
      if (frame.type === 'stdin_ack' || frame.type === 'close_stdin_ack') {
        this.resolveControlAck(String(frame.type))
        return
      }
      if (frame.type === 'error') {
        this.finish(new SandboxError(String(frame.message ?? frame.code ?? 'process error')))
        return
      }
      const waiter = this.waiters.shift()
      if (waiter) waiter({ done: false, value: frame })
      else this.queue.push(frame)
    } catch (error) {
      this.finish(error instanceof Error ? error : new SandboxError(String(error)))
    }
  }

  private finish(error?: Error): void {
    this.closed = true
    if (this.keepalive) clearInterval(this.keepalive)
    if (error) {
      const frame = { type: 'error', message: error.message }
      const waiter = this.waiters.shift()
      if (waiter) waiter({ done: false, value: frame })
      else this.queue.push(frame)
    }
    this.rejectControlAcks(error ?? new SandboxError('process websocket closed before acknowledgement'))
    this.flushDone()
  }

  private flushDone(): void {
    for (const waiter of this.waiters.splice(0)) {
      waiter({ done: true, value: undefined })
    }
  }

  private waitForControlAck(type: string): { promise: Promise<void>; cancel: () => void } {
    let entry: { resolve: () => void; reject: (error: Error) => void; timer: ReturnType<typeof setTimeout> }
    const promise = new Promise<void>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.removeControlAck(type, entry)
        reject(new TimeoutError())
      }, this.requestTimeoutMs)
      entry = { resolve, reject, timer }
      const waiters = this.ackWaiters.get(type) ?? []
      waiters.push(entry)
      this.ackWaiters.set(type, waiters)
    })

    return {
      promise,
      cancel: () => {
        this.removeControlAck(type, entry)
        clearTimeout(entry.timer)
      },
    }
  }

  private resolveControlAck(type: string): void {
    const entry = this.ackWaiters.get(type)?.shift()
    if (!entry) return
    clearTimeout(entry.timer)
    entry.resolve()
  }

  private rejectControlAcks(error: Error): void {
    for (const waiters of this.ackWaiters.values()) {
      for (const entry of waiters.splice(0)) {
        clearTimeout(entry.timer)
        entry.reject(error)
      }
    }
    this.ackWaiters.clear()
  }

  private removeControlAck(type: string, entry: { timer: ReturnType<typeof setTimeout> }): void {
    const waiters = this.ackWaiters.get(type)
    if (!waiters) return
    const index = waiters.indexOf(entry as { resolve: () => void; reject: (error: Error) => void; timer: ReturnType<typeof setTimeout> })
    if (index !== -1) waiters.splice(index, 1)
    if (waiters.length === 0) this.ackWaiters.delete(type)
  }
}

export function base64Encode(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString('base64')
}

export function base64DecodeText(value: unknown): string {
  if (typeof value !== 'string') return String(value ?? '')
  try {
    return Buffer.from(base64DecodeBytes(value)).toString('utf8')
  } catch {
    return value
  }
}

export function base64DecodeBytes(value: unknown): Uint8Array {
  if (typeof value !== 'string') return new TextEncoder().encode(String(value ?? ''))
  return new Uint8Array(Buffer.from(value, 'base64'))
}

function rawDataToText(message: RawData): string {
  if (typeof message === 'string') return message
  if (Array.isArray(message)) return Buffer.concat(message).toString('utf8')
  if (message instanceof ArrayBuffer) return Buffer.from(new Uint8Array(message)).toString('utf8')
  return Buffer.from(message as Uint8Array).toString('utf8')
}

function wsUrl(baseUrl: string, path: string): string {
  const url = new URL(path, baseUrl.replace(/^http/, 'ws'))
  return url.toString()
}
